Suppress pedantic diagnostic for a file not ending in EOL

AaronBallman opened the pull request:
WG14 added N3411 to the list of papers which apply to older versions of C in C2y, and WG21 adopted CWG787 as a Defect Report in C++11. So we no longer should be issuing a pedantic diagnostic about a file which does not end with a newline character.

We do, however, continue to support -Wnewline-eof as an opt-in diagnostic.

AaronBallman committed:
Suppress pedantic diagnostic for a file not ending in EOL

WG14 added N3411 to the list of papers which apply to older versions of
C, and WG21 adopted CWG787 as a Defect Report in C++11. So we no longer
should be issuing a pedantic diagnostic about a file which does not end
with a newline character.

We do, however, continue to support -Wnewline-eof as an opt-in
diagnostic.

cor3ntin commented on clang/test/Lexer/newline-eof.c:
```diff
@@ -1,11 +1,16 @@
-// RUN: %clang_cc1 -fsyntax-only -x c++ -std=c++03 -pedantic -verify %s
-// RUN: %clang_cc1 -fsyntax-only -Wnewline-eof %s 2>&1 | FileCheck %s
+// Allowing a file to end without a newline was adopted as a Defect Report in
+// WG21 (CWG787) and in WG14 (added to the list of changes which apply to
```
Can you add a test in the appropriate c++ dr test file ?
I'm happy  with just adding // dr787 : yes + a comment pointing to this test file
(+ rebuilding cxx_dr_status)

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -128,7 +128,10 @@ C2y Feature Support
 - Implemented `WG14 N3411 <https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3411.pdf>`_
   which allows a source file to not end with a newline character. This is still
-  reported as a conforming extension in earlier language modes.
+  reported as a conforming extension in earlier language modes. Note,
```
I thought this is removing the 'reported as a conforming extension in earlier language modes'?  Or is my lack of caffeine showing?

AaronBallman replied:
Nope, that's a think-o on my part. :-D

erichkeane commented on clang/lib/Lex/Lexer.cpp:
```diff
@@ -3194,18 +3194,12 @@ bool Lexer::LexEndOfFile(Token &Result, const char *CurPtr) {
+    // C++11 [lex.phases] 2.2 p2
+    // Prefer the C++98 pedantic compatibility warning over the generic,
+    // non-extension, user-requested "missing newline at EOF" warning.
+    if (LangOpts.CPlusPlus11 &&
```
It isn't clear what the status of this in C++98 actually is, right?  I guess this is before WG21 was all crazy about everything-as-a-DR so we don't have much guidance here from them, but it feels DR-ish for both.

AaronBallman replied:
It is a DR in C++11 so it applies to C++98 as well. What this block is doing is basically saying "if the user passed `-Wc++98-compat-pedantic` prefer issuing that diagnostic over the `-Wnewline-eof` diagnostic."

erichkeane replied:
Yeah, I get that.  But if this is a DR, it ALSO applies to C++98, so it isn't a compatibility warning in C++98.

AaronBallman committed:
Add C++ DR test

Note: I wasn't able to run the python script to regenerate the
cxx_dr_status.html page due to encoding errors, so that still needs to
be run.

AaronBallman committed:
Update release notes

AaronBallman committed:
Remove the C++98 compat diagnostic

Review comments pointed out that this is no longer a necessary
diagnostic given that C++11 adopted the change as a defect report.

Endilll commented on clang/test/CXX/drs/cwg787.cpp:
```diff
@@ -0,0 +1,13 @@
+// This file intentionally does not end with a newline. CWG787 made this
+// well-defined behavior.
+
+// cwg787: 3.0
```
How can we claim that Clang 3.0 is conforming when you are fixing this right in this PR? https://godbolt.org/z/1rTq148Wd

AaronBallman replied:
Because we accept the code and do the right thing with it? AIUI, the Core Issue was that it went from UB to well-defined behavior, but we diagnose well-defined behavior all the time; it doesn't mean we don't implement the DR, right? (I'm fine with changing the marking, it just wasn't clear to me that any changes were needed to conform to the DR.)

Endilll replied:
> Because we accept the code and do the right thing with it?

`-pedantic-errors` reject it up until your patch, IIUC.

> AIUI, the Core Issue was that it went from UB to well-defined behavior, but we diagnose well-defined behavior all the time; it doesn't mean we don't implement the DR, right? (I'm fine with changing the marking, it just wasn't clear to me that any changes were needed to conform to the DR.)

My reading of changes to [lex.phases] in [CWG787](https://cplusplus.github.io/CWG/issues/787.html) doesn't leave a room for issuing a diagnostic while remaining conformant. Maybe this is the source of the confusion. When we determine availability of a DR in older version, we check them against our current understanding of what the behavior should be. In this case, I think it is that no diagnostic should be issued. Which means `// cwg787: 21` is the correct status here.

Let me know if I'm missing something.

AaronBallman replied:
Okay, that works for me, thanks!

AaronBallman committed:
Fix build error

Endilll reviewed (commented):
`cxx_dr_status.html` entry should be updated, too.
It will be trivial to do once #131812 is merged.

AaronBallman committed:
Update DR test based on review feedback

zygoloid commented on clang-tools-extra/clangd/Preamble.cpp:
```diff
@@ -627,8 +627,6 @@ buildPreamble(PathRef FileName, CompilerInvocation CI,
-    case diag::warn_cxx98_compat_no_newline_eof:
-    case diag::ext_no_newline_eof:
       // If the preamble doesn't span the whole file, drop the no newline at
       // eof warnings.
```
```suggestion
      // eof warning.
```

zygoloid commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -127,8 +127,10 @@ C2y Feature Support
-  which allows a source file to not end with a newline character. This is still
-  reported as a conforming extension in earlier language modes.
+  which allows a source file to not end with a newline character. Note,
+  ``-pedantic`` will no longer diagnose this in either C or C++ modes as an
```
This sounds to me like you're saying that not diagnosing this is an extension. Maybe drop the "as an extension" part?

zygoloid reviewed (commented):
LG

AaronBallman committed:
Update based on review feedback

Endilll commented:
Both #131812 and #131816 have been merged, so you can update your branch and run the script.

AaronBallman commented:
> Both #131812 and #131816 have been merged, so you can update your branch and run the script now.

Thank you! I re-ran the script and it seems to have an issue with line endings because the diff shows changes on every line in the file. I'm going to land these changes as-is, the script can be fixed up and list regenerated in a follow-up.
